fedora-ostree-pruner/fedora-ostree-pruner

```diff
@@ -0,0 +1,261 @@
+                               (OSTREEPRODREPO, ref))
+            continue
+
+    # prune each branch in the policy with specified value
+    for ref,policy in PROD_REF_POLICIES.items():
+
+        # Skip if there is no ref in the repo for this policy
+        if ref not in prod_refs:
+            continue
+
+        if policy is None:
+            logger.info('Skipping ref %s in repo %s. Policy is to keep all commits' %
+                        (ref, OSTREEPRODREPO))
+            continue
+
+        if policy == 'delete':
+            delete_ref_in_repo(ref, OSTREEPRODREPO, test)
+            continue
+
+        prunearg = parse_policy(policy)
+
+        logger.info('Pruning the %s ref in repo %s to %s' %
+                    (ref, OSTREEPRODREPO, policy))
+        cmd = ['ostree', 'prune', '--repo', OSTREEPRODREPO,
+               '--only-branch', ref, '--refs-only', prunearg]
+        if test:
+            cmd.append('--no-prune')
+        runcmd(cmd, capture_output=False)
+
+
+# XXX this function should not be used yet as there are a few
+# bugs to work out: 
+# https://github.com/ostreedev/ostree/issues/1479 
+# https://github.com/ostreedev/ostree/issues/1481
+def prune_prod_repo_deltas(test=False):
+
+    logger.info("prune_prod_repo_deltas: this function is disabled until the"
+                "following bugs are fixed:")
+    logger.info("https://github.com/ostreedev/ostree/issues/1479")
+    logger.info("https://github.com/ostreedev/ostree/issues/1481")
+    return
+
+    cmd = ['ostree', 'prune', '--repo', OSTREEPRODREPO,
+           '--refs-only', '--static-deltas-only']
+    if test:
+        cmd.append('--no-prune')
+    runcmd(cmd, capture_output=False)
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--test",
+                        help="Don't actually prune", action='store_true')
+    parser.add_argument("--loop",
+                        help="Loop forever, run once a day.", action='store_true')
+    args = parser.parse_args()
+
+    # set args.test = True for now. Once we have proved
+    # everything out in prod and the logs look good we'll
+    # delete this code.
+    args.test = True
+
+    # on startup let's print out the config that was generated
+    logger.info('The configured policy is: \n%s' %
+            pprint.pformat(PROD_REF_POLICIES, indent=8))
+
+    # Configure a umask of 0002 which will allow for the group permissions
+    # to include write for newly created files. We need this because we'd
+    # like to access the OSTree repos from two different Kubernetes pods,
+    # which will have different UIDs but the same GID.
+    # See https://pagure.io/releng/issue/8811#comment-616490
+    os.umask(0o0002)
+
+    # Prune the compose and the prod repo based on policy
+    prune_compose_repo(args.test)
+    prune_prod_repo(args.test)
+
+    # If we were asked to run in a loop, then run once a day
+    if args.loop:
```

Comment:
the reason I wanted to run this in a loop rather than run it as part of a kubernetes cron jobs is because historically it's been really hard to get logs from openshift pods in fedora infra. I've recently learned we have a kibana instance and should be able to get logs more easily. I'll try to confirm that and then maybe we'll convert this into a cronjob..

on the topic of frequency. Pruning is actually pretty I/O intensive (since it has to read a lot from the repo just to see if there is anything to prune) and slow on these large repos. I think maybe we should just run it on like Friday or Saturday (once a week) during the night.